packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx

```diff
@@ -41,47 +42,64 @@ function addLabelDimension(
     tickLabelInterval,
     reverse,
     isMounted,
+    isPointInside,
   }: Pick<ChartsXAxisProps, 'tickLabelInterval' | 'tickLabelStyle'> &
-    Pick<AxisDefaultized, 'reverse'> & { isMounted: boolean },
+    Pick<AxisDefaultized, 'reverse'> & {
+      isMounted: boolean;
+      isPointInside: ChartInstance<unknown>['isPointInside'];
+    },
 ): (TickItemType & LabelExtraData)[] {
-  const withDimension = xTicks.map((tick) => {
+  const getTickSize = (tick: TickItemType) => {
     if (!isMounted || tick.formattedValue === undefined) {
-      return { ...tick, width: 0, height: 0 };
+      return { width: 0, height: 0 };
     }
+
     const tickSizes = getWordsByLines({ style, needsComputation: true, text: tick.formattedValue });
+
     return {
-      ...tick,
       width: Math.max(...tickSizes.map((size) => size.width)),
       height: Math.max(tickSizes.length * tickSizes[0].height),
     };
-  });
+  };
 
-  if (typeof tickLabelInterval === 'function') {
-    return withDimension.map((item, index) => ({
-      ...item,
-      skipLabel: !tickLabelInterval(item.value, index),
-    }));
-  }
+  // FIXME: Add this back
+  // if (typeof tickLabelInterval === 'function') {
+  //  return withDimension.map((item, index) => ({
+  //    ...item,
+  //    skipLabel: !tickLabelInterval(item.value, index),
+  //  }));
+  // }
 
   // Filter label to avoid overlap
-  let currentTextLimit = 0;
   let previousTextLimit = 0;
   const direction = reverse ? -1 : 1;
-  return withDimension.map((item, labelIndex) => {
-    const { width, offset, labelOffset, height } = item;
+  const minGap = 8;
 
-    const distance = getMinXTranslation(width, height, style?.angle);
+  return xTicks.map((item, labelIndex) => {
+    const { offset, labelOffset } = item;
     const textPosition = offset + labelOffset;
-    const gapRatio = 1.2; // Ratio applied to the minimal distance to add some margin.
 
-    currentTextLimit = textPosition - (direction * (gapRatio * distance)) / 2;
-    if (labelIndex > 0 && direction * currentTextLimit < direction * previousTextLimit) {
+    if (labelIndex > 0 && direction * (textPosition - minGap) < direction * previousTextLimit) {
+      return { ...item, width: 0, height: 0, skipLabel: true };
+    }
+
+    if (!isPointInside({ x: textPosition, y: -1 }, { direction: 'x' })) {
+      return { ...item, width: 0, height: 0, skipLabel: true };
+    }
```

Comment:
Don't calculate label width if the tick is outside the visible area (useful when zooming)